File: packages/typescript/examples/select-npcs-by-pole.ts
import {
  AdaptiveBifurcatedDomain,
  type AdaptiveBifurcatedDomainValue,
  AdaptiveBifurcatedPole,
  type AdaptiveBifurcatedPoleValue,
  AdaptiveBifurcatedProfile,
} from "../dist/index.js";

const profile = (
  vector: readonly [number, number, number, number, number],
): AdaptiveBifurcatedProfile =>
  new AdaptiveBifurcatedProfile({
    orderScore: vector[0],
    chaosScore: vector[1],
    cooperationScore: vector[2],
    conflictScore: vector[3],
    competitionScore: vector[4],
  });

const cast = {
  Asha: profile([0.7, 0.5, 0.8, 0.4, 0.3]),
  Bram: profile([-0.4, -0.7, -0.6, -0.5, -0.8]),
  Cato: profile([0.2, 0.1, -0.2, -0.7, 0.8]),
  Dima: profile([-0.6, 0.6, 0.5, 0.7, 0.1]),
};

function selectByPole(
  domain: AdaptiveBifurcatedDomainValue,
  pole: AdaptiveBifurcatedPoleValue,
  threshold = 0.4,
): string[] {
  return Object.entries(cast)
    .filter(([, candidate]) => candidate.dominantPoles(threshold)[domain] === pole)
    .map(([name]) => name);
}

console.log(
  "Collaborative NPCs:",
  selectByPole(
    AdaptiveBifurcatedDomain.Cooperation,
    AdaptiveBifurcatedPole.Collaboration,
  ),
);
console.log(
  "Dominancy-leaning NPCs:",
  selectByPole(AdaptiveBifurcatedDomain.Competition, AdaptiveBifurcatedPole.Dominancy),
);

File: packages/python/examples/select_npcs_by_pole.py
"""Filter NPCs by dominant ABBF poles."""

from __future__ import annotations

from personalitygen import (
    AdaptiveBifurcatedDomain,
    AdaptiveBifurcatedPole,
    AdaptiveBifurcatedProfile,
)


CAST = {
    "Asha": AdaptiveBifurcatedProfile(0.7, 0.5, 0.8, 0.4, 0.3),
    "Bram": AdaptiveBifurcatedProfile(-0.4, -0.7, -0.6, -0.5, -0.8),
    "Cato": AdaptiveBifurcatedProfile(0.2, 0.1, -0.2, -0.7, 0.8),
    "Dima": AdaptiveBifurcatedProfile(-0.6, 0.6, 0.5, 0.7, 0.1),
}


def select_by_pole(
    domain: AdaptiveBifurcatedDomain,
    pole: AdaptiveBifurcatedPole,
    *,
    threshold: float = 0.4,
) -> list[str]:
    matches = []
    for name, profile in CAST.items():
        poles = profile.dominant_poles(threshold=threshold)
        if poles.get(domain) == pole:
            matches.append(name)
    return matches


def main() -> None:
    collaborators = select_by_pole(
        AdaptiveBifurcatedDomain.COOPERATION,
        AdaptiveBifurcatedPole.COLLABORATION,
    )
    dominancy = select_by_pole(
        AdaptiveBifurcatedDomain.COMPETITION,
        AdaptiveBifurcatedPole.DOMINANCY,
    )

    print(f"Collaborative NPCs: {', '.join(collaborators)}")
    print(f"Dominancy-leaning NPCs: {', '.join(dominancy)}")


if __name__ == "__main__":
    main()
